### DISCUSSION TITLE
Hierarchical GPs

### DISCUSSION THREAD

**Discussion by crimbs** [thomaspinder/GPJax#185] [Ideas]
Hi team. I'm looking to implement the GP clustering method described in [Hierarchical Bayesian modelling of gene expression time series across irregularly sampled replicates and clusters](https://doi.org/10.1186/1471-2105-14-252) using GPJax. 

Looking at GPy's implementation used in the paper as well as @thomaspinder's [implementation in in GPFlow](https://github.com/thomaspinder/HierarchicalGPFlow), it seems that the notion of a hierarchical kernel fits naturally into the `CombinationKernel` base class. However in GPJax this would require passing additional arguments to the pairwise `combination_fn` indicating which replicate the data point belongs to. For instance, see equation (5) in [Fast Nonparametric Clustering of Structured Time-Series](https://doi.org/10.1109/TPAMI.2014.2318711)

<img width="608" alt="Screenshot 2023-02-09 at 11 26 33" src="https://user-images.githubusercontent.com/48622823/217800115-c14efa4d-09c6-4db8-9318-2e08765e5a41.png">

The piecewise function above could then be written using [`lax.cond`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html) (or [`lax.switch`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.switch.html) for deeper hierarchies). 

Do you think this is a over-engineered solution to remain consistent with GPJax's `vmap`-over-pairwise-kernel-functions approach instead of defining the covariance matrix using something like
```python
def Ktilde(t, td, n, nd):
    K = kf.cross_covariance(kf_params, t, td)
    mask = n == nd.T
    K += mask * kg.cross_covariance(kg_params, t, td)
    return K
```
It would be nice to integrate smoothly with GPJax to experiment with sparse methods etc.

**Comment by thomaspinder** [comment]
Hi @crimbs, thanks for starting this discussion. You might be interested in the paper [Identifying latent climate signals using sparse
hierarchical Gaussian processes](https://gp-seminar-series.github.io/neurips-2022/assets/camera_ready/3.pdf) where do adapt the hierarchical GP to be amenable with inducing points. 

As you point out, implementing in GPJax is _different_ to how one might do it in GPFlow. However, using [kernex](https://github.com/ASEM000/kernex), I think this could lead to a performant implementation. In GPFlow, it is very hard to avoid a for-loop which yields a huge PyTree.

I'd be very happy to support this in JaxKern as I believe the hierarchical kernel to be incredibly useful.
